R/metab_simple.R

```diff
@@ -48,31 +54,39 @@ metab_simple <- function(data, ...) {
   #' 
   #' @param params a vector of length 2, where the first element is GPP and the second element is ER (both mg/L/d)
   #' @param 
-#   onestation_likelihood <- function(params, DO, DO.deficit, oxy, light, z, bp, ts, K) {
-#     
-#     GPP <- params[1]
-#     ER <- params[2]
-#     
-#     DO.modeled <- numeric(nrow(data))
-#     DO.modeled[1] <- oxy[1]
-#     for (i in 2:length(DO)) {
-#       DO.modeled[i] <- 
-#         DO.modeled[i-1] +
-#         (GPP/z)*(light[i]/sum(light)) + 
-#         ER * ts / z + 
-#         Kcor(temp[i],K)*ts*DO.deficit 
-#     }
-#   
-#     ##below is MLE calculation
-#     sqdiff<-(oxy-metab)^2 
-#     length(oxy)*(log(((sum(sqdiff)/length(oxy))^0.5)) +0.5*log(6.28))   + ((2*sum(sqdiff)/length(oxy))^-1)*sum(sqdiff)
-#   }
-  
+  onestation_negloglik <- function(params, DO.obs, DO.deficit, depth, k.O2, frac.GPP, frac.ER, frac.D) {
+    
+    # Parse params vector (passed from nlm)
+    GPP <- params[1]
+    ER <- params[2]
+    
+    # Model DO with given params
+    DO.mod <- numeric(length(DO.obs))
+    DO.mod[1] <- DO.obs[1]
+    DO.delta <- 
+      GPP * frac.GPP / depth + 
+      ER * frac.ER / depth + 
+      k.O2 * DO.deficit * frac.D # Bob - pros & cons of using DO.mod[t-1] instead of DO.obs[t] or DO.obs[t-1] (i.e., from DO.deficit) here?
+    DO.mod[-1] <- DO.mod[1] + cumsum(DO.delta[-1])
+    
+    # calculate & return the negative log likelihood of DO.mod values relative
+    # to DO.obs values. equivalent to Bob's original code & formula at
+    # http://www.statlect.com/normal_distribution_maximum_likelihood.htm
+    n <- length(DO.obs)
+    diffs.sq <- (DO.obs-DO.mod)^2 
+    sigma.sq <- sum(diffs.sq)/n
+    (n/2)*log(sigma.sq) + (n/2)*log(2*pi) + (1/(2*sigma.sq))*sum(diffs.sq)
+  }
   
   # Calculate metabolism by non linear minimization of an MLE function
-#   river.mle <- nlm(mlefunction, p=c(3,-5), oxy=oxy, z=z,temp=temp,light=light, bp=bp, ts=ts, K=K)
-  
-#   b<-c(river.mle$estimate[1], river.mle$estimate[2],river.mle$estimate[3],  river.mle$minimum[1])
+  river.mle <- with(
+    data, 
+    nlm(onestation_negloglik, p=c(GPP=3,ER=-5), 
+        DO.obs=DO.obs, DO.deficit=DO.deficit, depth=depth, k.O2=k.O2, 
+        frac.GPP=light/sum(light), frac.ER=timestep.days, frac.D=timestep.days)
+  )
```

Comment:
(It's probably easiest to view the entire file at https://github.com/aappling-usgs/streamMetabolizer/blob/master/R/metab_simple.R)